File: rust/arDEWino-rs/src/usb.rs
use std::error::Error;
use std::time::{Duration, Instant};

use regex::Regex;
use tokio::time::sleep;
use tokio_serial::{SerialStream, SerialPortBuilderExt};
use tokio::io::{AsyncWriteExt, AsyncReadExt};

use crate::models::IndoorSensorData;

pub struct UsbCommunication {
    port: SerialStream,
}

impl UsbCommunication {
    pub async fn new(port_name: &str) -> Result<Self, Box<dyn Error>> {
        let port = tokio_serial::new(port_name, 115200)
            .open_native_async()?;
        Ok(UsbCommunication { port })
    }

    async fn write_data(&mut self, data: &str) -> Result<(), Box<dyn Error>> {
        self.port.write_all(data.as_bytes()).await?;
        Ok(())
    }

    async fn read_data(&mut self) -> Result<String, Box<dyn Error>> {
        let mut buffer = [0u8; 32];
        let n = self.port.read(&mut buffer).await?;
        
        if n > 0 {
            let result = String::from_utf8_lossy(&buffer[..n]).to_string();
            Ok(result)
        } else {
            Ok(String::new())
        }
    }

    pub async fn get_arduino_response(&mut self, command: &str, sleep_duration: u64) -> Result<String, Box<dyn Error>> {
        let start_time = Instant::now();
        let max_duration = Duration::from_millis(1000);

        println!("Waiting for command '{}' ack...", command);

        loop {
            self.write_data(command).await?;
            let response = self.read_data().await?;
            let trimmed_response = response.trim();

            if is_valid_response(trimmed_response) {
                println!("Arduino says: {}", response);
                return Ok(response);
            }
            if start_time.elapsed() >= max_duration {
                return Err("Arduino not responding...".into());
            }
            sleep(Duration::from_millis(sleep_duration)).await;
        }
    }

    pub async fn get_indoor_sensor_data(&mut self) -> Result<IndoorSensorData, Box<dyn Error>> {
        let command = "d";
        let arduino_data = self.get_arduino_response(command, 50).await?;
        let parts: Vec<&str> = arduino_data.split(',').collect();
        
        let device_id: u64 = parts[0].trim().parse()?;
        let temperature: f32 = parts[1].trim().parse()?;
        let humidity: f32 = parts[2].trim().parse()?;
        let led_state: bool = match parts[3].trim() {
            "1" => true,
            "0" => false,
            _ => return Err("Invalid LED state value".into()),
        };
    
        Ok(IndoorSensorData {
            device_id,
            temperature,
            humidity,
            led_state
        })
    }

    pub async fn toggle_warning_light(&mut self, open_windows: bool) -> Result<(), Box<dyn Error>> {
        if open_windows {
            self.get_arduino_response("0", 50).await?;
            println!("Warning light OFF");
        } else {
            self.get_arduino_response("1", 50).await?;
            println!("Warning light ON");
        }
        Ok(())
    }
}

fn is_valid_response(response: &str) -> bool {
    !response.is_empty() && (is_valid_data_format(response) || response == "a")
}

pub fn is_valid_data_format(input: &str) -> bool {
    let pattern = Regex::new(r"^\d{1,20},\d{2}\.\d{2},\d{2}\.\d{2},[01]$").unwrap();
    pattern.is_match(input)
}

File: go/dewdrop-go/internal/usb/usb.go
package usb

import (
	"bytes"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/mugglemath/dewdrop-go/pkg/models"
	"github.com/mugglemath/dewdrop-go/pkg/utils"
	"github.com/tarm/serial"
)

type Client interface {
	GetIndoorSensorData() (models.IndoorSensorData, error)
	ToggleWarningLight(openWindows bool) error
}

type SerialPort interface {
	Write(data []byte) (int, error)
	Read(buffer []byte) (int, error)
}

type usbClientImpl struct {
	port SerialPort
}

func NewUsbCommunication(portName string) (*usbClientImpl, error) {
	c := &serial.Config{
		Name: portName,
		Baud: 115200,
	}
	port, err := serial.OpenPort(c)
	if err != nil {
		return nil, err
	}
	return &usbClientImpl{port: port}, nil
}

// GetIndoorSensorData retrieves the sensor data from the Arduino
func (usb *usbClientImpl) GetIndoorSensorData() (models.IndoorSensorData, error) {
	command := "d"
	response, err := usb.getArduinoResponse(command, 50*time.Millisecond)
	if err != nil {
		return models.IndoorSensorData{}, err
	}

	parts := utils.SplitAndTrim(response, ',')
	if len(parts) < 4 {
		return models.IndoorSensorData{}, errors.New("invalid data format")
	}

	deviceID, _ := strconv.ParseUint(parts[0], 10, 64)
	temperature, _ := strconv.ParseFloat(parts[1], 32)
	humidity, _ := strconv.ParseFloat(parts[2], 32)
	var ledState bool
	if parts[3] == "1" {
		ledState = true
	} else if parts[3] == "0" {
		ledState = false
	} else {
		return models.IndoorSensorData{}, errors.New("invalid LED state value")
	}

	return models.IndoorSensorData{
		DeviceID:    deviceID,
		Temperature: float32(temperature),
		Humidity:    float32(humidity),
		LedState:    ledState,
	}, nil
}

// ToggleWarningLight toggles the blinking yellow light on the Arduino
func (usb *usbClientImpl) ToggleWarningLight(openWindows bool) error {
	command := "1"
	if openWindows {
		command = "0"
	}

	_, err := usb.getArduinoResponse(command, 50*time.Millisecond)
	if err != nil {
		return err
	}

	if openWindows {
		fmt.Println("Warning light OFF")
	} else {
		fmt.Println("Warning light ON")
	}

	return nil
}

// getArduinoResponse sends the Arduino a command and retrieves the response
func (usb *usbClientImpl) getArduinoResponse(command string, sleepDuration time.Duration) (string, error) {
	startTime := time.Now()
	maxDuration := time.Millisecond * 500

	fmt.Printf("Waiting for command '%s' ack...\n", command)

	for {
		if err := usb.writeData(command); err != nil {
			return "", err
		}

		response, err := usb.readData()
		if err != nil {
			return "", err
		}

		if utils.IsValidResponse(response) {
			fmt.Printf("Arduino says: %s\n", response)
			return response, nil
		}

		if time.Since(startTime) >= maxDuration {
			return response, nil
		}

		time.Sleep(sleepDuration)
	}
}

func (usb *usbClientImpl) readData() (string, error) {
	buffer := make([]byte, 32)
	n, err := usb.port.Read(buffer)
	if err != nil {
		return "", err
	}
	if n > 0 {
		return string(bytes.TrimSpace(buffer[:n])), nil
	}
	return "", nil
}

func (usb *usbClientImpl) writeData(data string) error {
	_, err := usb.port.Write([]byte(data))
	return err
}
